src/bls/bls_aggregator.cpp

```diff
@@ -591,183 +663,262 @@ bls_rewards_response bls_aggregator::rewards_request(const address& addr, uint64
                 "403",
-                "Forbidden: The BLS pubkey {} is not currently removable."_format(
-                        removal_request.remove_pk));
+                "Forbidden: The BLS pubkey {} is not currently {}."_format(
+                        request.remove_pk, bls_exit_type_label(type)));
         return;
     }
 
     auto& signer = core.bls_signer();
 
-    std::vector<uint8_t> msg = get_removal_msg_to_sign(
+    std::vector<uint8_t> msg = get_exit_msg_to_sign(
             core.get_nettype(),
-            bls_aggregator::removal_type::normal,
-            removal_request.remove_pk,
-            removal_request.timestamp.count());
+            type,
+            request.remove_pk,
+            request.timestamp.count());
     bls_signature sig = signer.signMsg(msg);
 
     oxenc::bt_dict_producer d;
-    // BLS pubkey to remove:
-    d.append("remove", tools::view_guts(removal_request.remove_pk));
-    // signature of *this* snode of the removing pubkey:
-    d.append("signature", tools::view_guts(sig));
-
+    d.append("remove", tools::view_guts(request.remove_pk));  // BLS pubkey to remove
+    d.append("signature", tools::view_guts(sig));  // Signs over the exit key and timestamp
     m.send_reply("200", std::move(d).str());
 }
 
-void bls_aggregator::get_liquidation(oxenmq::Message& m) {
-    oxen::log::trace(logcat, "Received omq liquidation signature request");
-    bls_removal_request removal_request = extract_removal_request(m);
-    if (!removal_request.good)
-        return;
+// Common code for exit and liquidation requests, which only differ in three ways:
+// - the endpoint they go to;
+// - the tag that gets used in the msg_to_sign hash; and
+// - the key under which the signed pubkey gets confirmed back to us.
+bls_exit_liquidation_response bls_aggregator::exit_liquidation_request(
+        const crypto::public_key& pubkey, bls_exit_type type) {
+
+    // NOTE: Trace entry into function
+    std::string_view label = bls_exit_type_label(type);
+    oxen::log::trace(logcat, "Initiating {} request for SN {}", label, pubkey);
+
+    // NOTE: Lookup the BLS pubkey associated with the Ed25519 pubkey.
+    std::optional<eth::bls_public_key> maybe_bls_pubkey{};
+    for (const service_nodes::service_node_list::recently_removed_node& it :
+         core.service_node_list.recently_removed_nodes()) {
+        if (it.pubkey == pubkey) {
+            maybe_bls_pubkey = it.bls_pubkey;
+            break;
+        }
+    }
 
-    if (!core.is_node_liquidatable(removal_request.remove_pk)) {
-        m.send_reply(
-                "403",
-                "Forbidden: The BLS key {} is not currently liquidatable"_format(
-                        removal_request.remove_pk));
-        return;
+    if (!maybe_bls_pubkey) {
+        throw oxen::traced<std::invalid_argument>(
+                "{} request for SN {} at height {} is invalid, node is not in the list of recently exited nodes. Request rejected"_format(
```

Comment:
Formatting: (for some reason, clang-format won't touch these user-defined string literals)
```suggestion
                "{} request for SN {} at height {} is invalid, node is not in the list of recently"
                " exited nodes. Request rejected"_format(
```

Reply:
Done